File: app/src/main/java/com/buntagon/testarchitecturecomponents/data/util/BaseDataSource.java
package com.buntagon.testarchitecturecomponents.data.util;

import android.arch.lifecycle.LiveData;

import java.util.List;

import io.realm.RealmModel;

/**
 * Common functions exposed by data sources for a data model class
 * Created by Alex on 19/11/2017.
 */

public interface BaseDataSource<T extends RealmModel & SyncedItem> {

    void insertOrUpdate(T item);

    void insertOrUpdate(List<T> items);

    void delete(String id);

    LiveData<T> get(String id);

    LiveData<List<T>> getAll();
}

File: app/src/main/java/com/buntagon/testarchitecturecomponents/data/util/BaseDataSource.kt
package com.buntagon.testarchitecturecomponents.data.util

/**
 * Common functions exposed by data sources for a data model class
 * Created by Alex on 19/11/2017.
 */

interface BaseDataSource<T : SyncedItem> {

    fun getAll(): StaticLiveData<MutableList<T>>

    fun insertOrUpdate(item: T)

    fun insertOrUpdate(items: List<T>)

    fun delete(item: T)

    operator fun get(id: String): StaticLiveData<T>

    fun close()
}
